File: server/routes/files.js
const express = require("express");
const { authenticateTokenOrApiKey } = require("../middleware/auth");
const FileStorageService = require("../services/FileStorageService");

const router = express.Router();
const fileStorage = new FileStorageService();

// GET /api/files/:storageKey - Serve file from embedded storage
router.get("/:storageKey", authenticateTokenOrApiKey, async (req, res) => {
  try {
    const { storageKey } = req.params;
    
    // Only serve from embedded storage
    if (fileStorage.storageType !== 'embedded') {
      return res.status(400).json({ 
        error: "This endpoint only serves embedded storage files" 
      });
    }

    const { content, metadata } = await fileStorage.getFileContent(storageKey);

    // Set appropriate headers
    res.set({
      'Content-Type': metadata.mimeType,
      'Content-Length': content.length,
      'Content-Disposition': `attachment; filename="${metadata.originalName}"`,
      'Cache-Control': 'private, max-age=3600' // Cache for 1 hour
    });

    res.send(content);
  } catch (error) {
    console.error("File serving error:", error);
    
    if (error.message.includes('File not found')) {
      return res.status(404).json({ error: 'File not found' });
    }
    
    res.status(500).json({ error: 'Failed to serve file' });
  }
});

// GET /api/files/:storageKey/info - Get file metadata
router.get("/:storageKey/info", authenticateToken, async (req, res) => {
  try {
    const { storageKey } = req.params;
    
    if (fileStorage.storageType === 'embedded') {
      const { metadata } = await fileStorage.getFileContent(storageKey);
      
      res.json({
        success: true,
        metadata: {
          ...metadata,
          downloadUrl: fileStorage.generateEmbeddedDownloadUrl({ storageKey })
        }
      });
    } else {
      // For S3, we would need to store metadata separately or retrieve from S3
      res.status(501).json({ 
        error: "Metadata retrieval for S3 files not implemented in this endpoint" 
      });
    }
  } catch (error) {
    console.error("File metadata error:", error);
    
    if (error.message.includes('File not found')) {
      return res.status(404).json({ error: 'File not found' });
    }
    
    res.status(500).json({ error: 'Failed to get file metadata' });
  }
});

module.exports = router;

File: server-go/internal/http/handlers/files.go
package handlers

import (
	"fmt"
	"net/http"

	"github.com/gin-gonic/gin"

	"your.module/config-manager/internal/files"
	"your.module/config-manager/internal/logger"
)

type FilesHandler struct {
	storage *files.StorageManager
	log     *logger.Logger
}

func NewFilesHandler(storage *files.StorageManager, log *logger.Logger) *FilesHandler {
	return &FilesHandler{storage: storage, log: log}
}

func (h *FilesHandler) Register(r *gin.RouterGroup) {
	r.GET("/files/:storageKey", h.serveFile)
	r.GET("/files/:storageKey/info", h.fileInfo)
}

func (h *FilesHandler) serveFile(c *gin.Context) {
	storageKey := c.Param("storageKey")

	data, metadata, err := h.storage.GetFileContent(c.Request.Context(), storageKey)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "File not found"})
		return
	}

	if metadata.StorageType != files.StorageEmbedded {
		c.JSON(http.StatusBadRequest, gin.H{"error": "This endpoint only serves embedded storage files"})
		return
	}

	c.Header("Content-Type", metadata.MimeType)
	c.Header("Content-Length", fmt.Sprintf("%d", metadata.Size))
	c.Header("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, metadata.OriginalName))
	c.Header("Cache-Control", "private, max-age=3600")
	c.Data(http.StatusOK, metadata.MimeType, data)
}

func (h *FilesHandler) fileInfo(c *gin.Context) {
	storageKey := c.Param("storageKey")

	_, metadata, err := h.storage.GetFileContent(c.Request.Context(), storageKey)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "File not found"})
		return
	}

	url, _ := h.storage.GenerateDownloadURL(c.Request.Context(), metadata)

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"metadata": map[string]any{
			"fileId":       metadata.FileID,
			"originalName": metadata.OriginalName,
			"mimeType":     metadata.MimeType,
			"size":         metadata.Size,
			"storageType":  metadata.StorageType,
			"storageKey":   metadata.StorageKey,
			"uploadDate":   metadata.UploadDate,
			"downloadUrl":  url,
		},
	})
}
